### PR TITLE
Enable minUTxOValue: 1000000 in CI for shelley

### PR DESCRIPTION
**Pull request by piotr-iohk**
# Issue Number

https://github.com/input-output-hk/cardano-wallet/issues/1914
https://github.com/input-output-hk/cardano-wallet/issues/1926
# Overview
- 08ba84e6c5b3fef96319ec777d7d9874e52280a1
  Enable minUTxOValue: 1000000 in CI for shelley
  
- ba675b0f2b011f19115e471b163b4ccaa91ac40f
  Adjust tests for transactions and addresses
  
- b815bef775ceaa6fbb0d201703fdbbe1ab86da61
  Introduce minUTxOValue named constant
  
- c0c146184dcc36b27136ab85c7b1c8937b6c807b
  Test for error message on spending minUtxOValue
  
- 7e45643f6f9802d7179d3868b76aa6c990b2a470
  Make dust Shelley wallets compy with minUTxOValue
  
- 711195d5ecd4e461fe64ba06a722b516fc6b87c6
  Adjust new tests to minUTxOValue
  
- 940a5e77763457f17a715a961176f1bff10138b5
  MIR wallets and minor adjustments
  
- ad64d0d9db33e2bb5e404cfc2793ec2cd8e66409
  Add top level describe for each integration test suite
  
- 40e7a51f4475e7fc7c8949edb9c910f3c21bba6e
  Make keyDeposit = 1000000
  
- 6f200817e654d8c9f054136acadce4e8bbf2bbb9
  Test adjustments
  
- 17d38be229dcde5f9de47416809a1150fd52f73d
  remove SHELLEY_CALCULATE_02 as it is already covered in SHELLEY_MIGRATE_02 (fee estimation is there also)
  
- 84f8ab3dc639b89a51b3a57028e5be3b57bf9d89
  Adjust STAKE_POOLS_JOIN_01rewards
  
- 35d0e50d80233b032a1019699a82de73cae6e605
  make ADDRESS_IMPORT_05 not flaky by adding eventually

# Comments

### CODE REVIEW

File: lib/core-integration/src/Test/Integration/Scenario/API/Shelley/Addresses.hs
Comment (jonathanknowles) [inline lib/core-integration/src/Test/Integration/Scenario/API/Shelley/Addresses.hs]: Suggestion: use `NumericUnderscores` here too:
```suggestion
                            "quantity": 1_000_000,
```

Did I mention that I love this extension? :smile_cat:

File: lib/core-integration/src/Test/Integration/Scenario/API/Shelley/Transactions.hs
Comment (jonathanknowles) [inline lib/core-integration/src/Test/Integration/Scenario/API/Shelley/Transactions.hs]:
```suggestion
                        "quantity": 1_000_000,
```

File: lib/core-integration/src/Test/Integration/Scenario/API/Shelley/Transactions.hs
Comment (jonathanknowles) [inline lib/core-integration/src/Test/Integration/Scenario/API/Shelley/Transactions.hs]:
```suggestion
                        "quantity": 2_000_000,
```

File: lib/core-integration/src/Test/Integration/Scenario/API/Shelley/Transactions.hs
Comment (jonathanknowles) [inline lib/core-integration/src/Test/Integration/Scenario/API/Shelley/Transactions.hs]:
```suggestion
                        "quantity": 1_000_000,
```

**Review by jonathanknowles — COMMENTED** [review]
Hi @piotr-iohk 

Looks good to me, though I have one suggestion: 

**Suggestion**: define a named constant `minUTxOValue` and use this constant in places where we have hard-coded the value of `1_000_000`.

**Justification**:

* Hard-coded values increase the effort required to read and understand the code. It's not always obvious to the uninitiated reader why this value or that value was used, or whether the choice of value is deliberate or arbitrary.
* Using a named constant `minUTxOValue` gives these values a clear meaning, a single place to add documentation, and makes it easier for us to make changes that affect this value in future. (Of course, we could search for `1_000_000` or `1000000`, but this number is _rather_ common, and _might_ have been used for other reasons.)
 
```hs
minUTxOValue :: Natural
minUTxOValue = 1_000_000
```

**Comment by piotr-iohk** [comment]
I have introduced `minUTxOValue` and use numeric underscores.

**Comment by piotr-iohk** [comment]
bors try

**Comment by piotr-iohk** [comment]
bors try

**Comment by piotr-iohk** [comment]
bors try

**Comment by piotr-iohk** [comment]
bors try

**Comment by piotr-iohk** [comment]
Some unit test failures... on mac:
```
Failures:

  test/unit/Network/Wai/Middleware/LoggingSpec.hs:119:5:
  1) Network.Wai.Middleware.Logging, Logging Middleware, GET, 200, no query
       uncaught exception: HttpException
       HttpExceptionRequest Request {
         host                 = "localhost"
         port                 = 53169
         secure               = False
         requestHeaders       = []
         path                 = "/get"
         queryString          = ""
         method               = "GET"
         proxy                = Nothing
         rawBody              = False
         redirectCount        = 10
         responseTimeout      = ResponseTimeoutDefault
         requestVersion       = HTTP/1.1
       }
        ConnectionTimeout

  To rerun use: --match "/Network.Wai.Middleware.Logging/Logging Middleware/GET, 200, no query/"
```

**Review by KtorZ — CHANGES_REQUESTED** [review]
Please do not merge this PR before #1909.

The calculation of fees in integration scenarios is a pain in the a**. This completely changes the behavior regarding small amount so a lot of edge-case scenarios are going to be a mess to update. I'd prefer dealing with this once #1909 has been merged to not add even more complexity to an already every growing PR.

**Comment by piotr-iohk** [comment]
bors try

**Comment by piotr-iohk** [comment]
I have rebased, but there are errors on submitting faucet funds after hard fork due to `minUTxOValue = 1_000_000`:

```
  Error while submitting tx: ApplyTxError [LedgerFailure (UtxowFailure (UtxoFailure (OutputTooSmallUTxO [TxOutCompact "a\SYN\SOH\DC2\DC1\238\t\239\164\194\235\tC\193B\198}z\219\164\EM\171v\166\236\238{\225\DC1" 10,TxOutCompact "a.K}\235#\246\DC1\ENQ\246\206\139t\178\163>\181\CAN:\186;\SOHZ).z\243a\142" 3,TxOutCompact "a/\vt\135\194\255e\243E\213<\249s\255\157\NAKK\191\233f\169\186\161^\196\160\232\ENQ" 2,TxOutCompact "a9\185s \218F\162\180Mw\234\145\&0z\128\220\143\243\185\187\148\212g\195J\223\184\210" 5,TxOutCompact "aG\129\172X\143\175b\170\221\160\165\132\228\ESC\249wn5\248:\240\128\128\r\142\149\ENQ\224" 1,TxOutCompact "aM\205\251'\210\199j\183hq\185\DLE\246\DEL\157\210AQ\171\230\239\235\160\254\161\187\165U" 3,TxOutCompact "a\149\149\166\225\&2\133\148$P\250\190\165\194\134\208\237\143\&5\206\189\140\194M\DC38P\162\226" 1,TxOutCompact "a\190$\210\170\242\234\241\160\239_\145\147\&9\164\142\135A\141\134\SO\195H\180\162\255X\191\&0" 1,TxOutCompact "a\191\216A\180\221\EOT\175S\194\ENQ\164VA$\EOT\228\DC1\ETX\190*\211\143\144=\167\214\229\163" 12,TxOutCompact "a\250r9l\ETBA\191\STX9C\EOTLv\ACK\156\214Z\165\DLEec\164\SUB\132\209r*\183" 5])))]
```

I suppose that if we want minUtXOValue in CI it needs to be updated via protocol update after that instead of being defined in `shelley-genesis.yaml.`

**Comment by KtorZ** [comment]
@piotr-iohk I believe the problem is one of the faucet wallet that we create explicitly full of Dust. Now `Dust` has a whole new meaning. I imagine two ways out of this:

a) We remove the special "dust wallet" from the **Shelley** wallets and only test it from **Byron**. We can set dust genesis funds for Byron addresses, so that should work. But for Shelley, we are transferring funds as soon as the hard fork happens so that indeed needs to abide by the minUTxO rule. In the end, testing the migration of dust from both wallets is sort of redundant since that is _exactly the same logic_ being called in both cases. We can still have migration from Shelley but simply, not with Dust. 

b) We could indeed issue a protocol update, but then test scenarios become dependent on the order they're executed (unless you "undo" the update after each scenario testing the `minUTxO`).

**Comment by piotr-iohk** [comment]
Ok, it makes sense to go with your proposition a, i.e. removing shelley dust wallet.

For b, I was thinking actually to do it only once in " test set up" but after those small transactions are being triggered.

**Comment by piotr-iohk** [comment]
bors try

**Comment by piotr-iohk** [comment]
bors try

**Comment by piotr-iohk** [comment]
bors try

**Comment by piotr-iohk** [comment]
bors try

**Comment by piotr-iohk** [comment]
Two failures from Hydra: :thinking: 
```
Failures:

  src/Test/Integration/Scenario/API/Shelley/Migrations.hs:144:17: 
  1) API Specifications, SHELLEY_MIGRATIONS, SHELLEY_CALCULATE_02 - Can calculate fee for wallet with dust (that complies with minUTxOValue).
       expected: Status {statusCode = 200, statusMessage = "OK"}
        but got: Status {statusCode = 403, statusMessage = "Forbidden"}
       
       from the following response: Left (DecodeFailure "Error in $: parsing Cardano.Wallet.Api.Types.ApiWalletMigrationInfo(ApiWalletMigrationInfo) failed, key \"migration_cost\" not found: Response {responseStatus = Status {statusCode = 403, statusMessage = \"Forbidden\"}, responseVersion = HTTP/1.1, responseHeaders = [(\"Transfer-Encoding\",\"chunked\"),(\"Date\",\"Wed, 09 Sep 2020 14:08:35 GMT\"),(\"Server\",\"Warp/3.3.5\")], responseBody = \"{\\\"code\\\":\\\"nothing_to_migrate\\\",\\\"message\\\":\\\"I can't migrate the wallet with the given id: 60bb5513e4e262e445cf203db9cf73ba925064d2, because it's either empty or full of small coins which wouldn't be worth migrating.\\\"}\", responseCookieJar = CJ {expose = []}, responseClose' = ResponseClose}")

  To rerun use: --match "/API Specifications/SHELLEY_MIGRATIONS/SHELLEY_CALCULATE_02 - Cannot calculate fee for wallet with dust (that complies with minUTxOValue)./"

  src/Test/Integration/Scenario/API/Shelley/StakePools.hs:153:5: 
  2) API Specifications, SHELLEY_STAKE_POOLS, STAKE_POOLS_JOIN_01 - Can join a pool, earn rewards and collect them
       uncaught exception: IOException of type UserError
       user error (Waited longer than 90s (more than 2 epochs) for an action to resolve. Action: "Certificates are inserted after quiting a pool". Error condition: Just (HUnitFailure (Just (SrcLoc {srcLocPackage = "cardano-wallet-core-integration-2020.8.3-GVIkg2PtT0PGByZc4fKRj", srcLocModule = "Test.Integration.Scenario.API.Shelley.StakePools", srcLocFile = "src/Test/Integration/Scenario/API/Shelley/StakePools.hs", srcLocStartLine = 297, srcLocStartCol = 46, srcLocEndLine = 297, srcLocEndCol = 65})) (ExpectedButGot Nothing "Outgoing" "Incoming")))
```

Weird, because both passed for me locally. In particular `SHELLEY_CALCULATE_02` has a "counterpart" `SHELLEY_MIGRATE_02` which actually migrates the same wallet... and that one passed...

**Comment by piotr-iohk** [comment]
bors try

**Comment by piotr-iohk** [comment]
bors try

**Comment by piotr-iohk** [comment]
I think I managed to adjust all tests to minUtxOValue now.
cc: @KtorZ @jonathanknowles

File: lib/core-integration/src/Test/Integration/Faucet.hs
Comment (KtorZ) [inline lib/core-integration/src/Test/Integration/Faucet.hs]: That's some interesting choice of indentation we have here.

File: lib/core-integration/src/Test/Integration/Scenario/API/Shelley/Migrations.hs
Comment (KtorZ) [inline lib/core-integration/src/Test/Integration/Scenario/API/Shelley/Migrations.hs:258]: Not so dusty anymore :sweat_smile: ... Does this test still has meaning? (we do test similar wallet migrations from other scenarios, don't we?)

File: lib/core-integration/src/Test/Integration/Scenario/API/Shelley/StakePools.hs
Comment (KtorZ) [inline lib/core-integration/src/Test/Integration/Scenario/API/Shelley/StakePools.hs:298]: Shouldn't this be the opposite? `quitFeeAmt - depositAmt` ?

File: lib/core-integration/src/Test/Integration/Scenario/API/Shelley/Transactions.hs
Comment (KtorZ) [inline lib/core-integration/src/Test/Integration/Scenario/API/Shelley/Transactions.hs:182]: :+1:

File: lib/core-integration/src/Test/Integration/Scenario/API/Shelley/StakePools.hs
Comment (piotr-iohk) [reply lib/core-integration/src/Test/Integration/Scenario/API/Shelley/StakePools.hs:298]: I changed depositAmt to be 1 000 000 lovelace (to satisfy minUTxO value), so it is bigger than the quit fee now...

File: lib/core-integration/src/Test/Integration/Faucet.hs
Comment (piotr-iohk) [reply lib/core-integration/src/Test/Integration/Faucet.hs]: :no_mouth:

File: lib/core-integration/src/Test/Integration/Scenario/API/Shelley/Migrations.hs
Comment (piotr-iohk) [reply lib/core-integration/src/Test/Integration/Scenario/API/Shelley/Migrations.hs:258]: > Not so dusty anymore

Indeed, however It's still sort of "dust" that complies with minUTxOValue I guess...

There are two of those tests now:
`SHELLEY_MIGRATE_02 - migrating wallet with 'dust' (that complies with minUTxOValue) should pass.` (this one)
`SHELLEY_MIGRATE_01_big_wallet`

where in both "dust" is now 1 ada... in the end I'd leave them both as they are still a bit different than just `fixtureWallet`...

File: lib/core-integration/src/Test/Integration/Scenario/API/Shelley/StakePools.hs
Comment (KtorZ) [reply lib/core-integration/src/Test/Integration/Scenario/API/Shelley/StakePools.hs:298]: Okay, my bad. That's **quitting**, so we are getting the deposit back. The API returns `deposit - fee`, the expression above translates to: `deposit - (deposit - fee)` = `fee`. OK.

File: lib/core-integration/src/Test/Integration/Faucet.hs
Comment (KtorZ) [inline lib/core-integration/src/Test/Integration/Faucet.hs]:
```suggestion
    , zip 
        (addresses $ SomeMnemonic onlyDustWallet) 
        (map Coin 
          [ 1_000_000
          , 1_000_000
          , 5_000_000
          , 12_000_000
          , 1_000_000
          , 5_000_000
          , 3_000_000
          , 10_000_000
          , 2_000_000
          , 3_000_000 
          ]
        )
```

**Review by KtorZ — APPROVED** [review]

**Pull request merged**
